openupgrade_scripts/scripts/mail/15.0.1.5/post-migration.py

```diff
@@ -64,6 +70,129 @@ def finish_migration_to_mail_group(env):
     )
 
 
+def _migrate_placeholder_char(string):
+    """
+    Replace dynamic placeholders in char/text fields:
+    Example: "Dear ${object.name}," -> "Dear {{object.name}},"
+    """
+    if not string:
+        return string
+    pattern = r"\$\{([^}]*)\}"
+    repl = r"{{\1}}"
+    return re.sub(pattern, repl, string)
+
+
+def _migrate_placeholder_html(string):
+    """
+    Replace dynamic placeholders in HTML fields:
+    Example: 'Your name is ${object.name}' -> 'Your name is <t t-out="object.name"></t>'
+    """
+    pattern = r"\$\{([^}]*)\}"
+
+    def repl(match):
+        (expression,) = match.groups()
+        expression = html.escape(expression)
+        return f'<t t-out="{expression}"></t>'
+
+    return re.sub(pattern, repl, string)
+
+
+def _migrate_end(string):
+    pattern = r"^%\s+end(if|for)\s*$"
+    repl = r"</t>"
+    return re.sub(pattern, repl, string, flags=re.MULTILINE)
+
+
+def _migrate_else(string):
+    pattern = r"^%\s+else\s*:?\s*$"
+    repl = r'<t t-else="">'
+    return re.sub(pattern, repl, string, flags=re.MULTILINE)
+
+
+def _migrate_if(string):
+    pattern = r"^%\s+if\s+(.+?)\s*:?\s*$"
+
+    def repl(match):
+        (if_expression,) = match.groups()
+        if_expression = html.escape(if_expression)
+        return f'<t t-if="{if_expression}">'
+
+    return re.sub(pattern, repl, string, flags=re.MULTILINE)
+
+
+def _migrate_for(string):
+    pattern = rf"^%\s+for\s+({_PY_VAR_PATTERN})\s+in\s+(.+?)\s*:?\s*$"
+
+    def repl(match):
+        var_name, loop_expression = match.groups()
+        loop_expression = html.escape(loop_expression)
+        return f'<t t-foreach="{loop_expression}" t-as="{var_name}">'
+
+    return re.sub(pattern, repl, string, flags=re.MULTILINE)
+
+
+def _migrate_set(string):
+    pattern = rf"^%\s+set\s+({_PY_VAR_PATTERN})\s*=\s*(.+?)\s*$"
+
+    def repl(match):
+        var_name, var_value = match.groups()
+        var_value = html.escape(var_value)
+        return f'<t t-set="{var_name}" t-value="{var_value}"/>'
+
+    return re.sub(pattern, repl, string, flags=re.MULTILINE)
+
+
+def _migrate_html_attributes(string):
+    soup = BeautifulSoup(string, multi_valued_attributes=None, features="html.parser")
```

Comment:
maybe you can use https://lxml.de/lxmlhtml.html here to avoid the dependency on bs?

Reply:
Yes, I think this can be done the same with standard Odoo XML manipulation library, which is lxml.